### PR TITLE
Add first class support for 'Editor-only' generated code

### PR DESCRIPTION
**Pull request by jamiebrynes7**
#### Description

In order to continue with #1375, I'll need to generate some code for rendering a given component. This code will be Editor only code. We currently have one 'Editor-only' generated assembly for the 'Build worker' menu items.

However, this was just sort of inserted into the existing generated code. As most generated code uses the namespace of the schema to derive its file path, this could cause collisions.

E.g.

```
package improbable.buildsystem;

...
```

Would generate code _inside_ the Build System 'Editor-only' assembly. 

This PR rectifies this and lays the groundwork for future 'Editor-only' generated code modules by:

- Adding a entirely separate directory for `Editor-only` generated code. (This can't go into source since otherwise this causes collisions). 
- Refactoring the `CodegenJob` classes slightly to allow for easy declaration as editor or non-editor. (It defaults to non-editor).

This is done when passing down arguments from your derived job class:

```csharp
public class MyCodegenJob : CodegenJob
{
	// Note the AsEditor() call.
	public MyCodegenJob(CodegenJobOptions opts, IFileSystem fs, DetailsStore store) : base (opts.AsEditor(), fs, store) {}
}
```

which will correctly set the base output directory path.

#### Tests

Verified that the build system generated code is placed in the correct place.

#### Documentation

- [x] Changelog
- [x] Upgrade guide (extra `.gitignore` entries)

### CODE REVIEW

File: test-project/.gitignore [hunk 1]
```diff
@@ -12,4 +12,6 @@ Assets/Generated/Source.meta
 Assets/Generated/Source/**/*.*
 !Assets/Generated/Source/improbable/tests/**/*.cs
 !Assets/Generated/Source/improbable/testschema/**/*.cs
-!Assets/Generated/Source/improbable/dependentschema/**/*.cs
\ No newline at end of file
+!Assets/Generated/Source/improbable/dependentschema/**/*.cs
+Assets/Generated/Editor.meta
+Assets/Generated/Editor/**/*.*
```
Comment (paulbalaji) [inline test-project/.gitignore:17]: newline

File: workers/unity/Packages/io.improbable.gdk.tools/GdkToolsConfigurationWindow.cs [hunk 2]
```diff
@@ -16,6 +16,7 @@ public class GdkToolsConfigurationWindow : EditorWindow
         internal const string VerboseLoggingLabel = "Verbose logging";
         internal const string CodegenLogOutputDirLabel = "Log output directory";
         internal const string CodegenOutputDirLabel = "C# output directory";
+        internal const string CodegenEditorOutputDirLabel = "C# Editor output directory";
```
Comment (paulbalaji) [inline workers/unity/Packages/io.improbable.gdk.tools/GdkToolsConfigurationWindow.cs:19]: As a drive-by, could update the label to make it clearer that this is for generated code

File: workers/unity/Packages/io.improbable.gdk.buildsystem/.codegen/Source/WorkerGenerationJob.cs [hunk 3]
```diff
@@ -18,8 +18,8 @@ public class WorkerGenerationJob : CodegenJob
         private readonly string relativeOutputPath = Path.Combine("improbable", "buildsystem");
         private readonly string relativeEditorPath = Path.Combine("improbable", "buildsystem", "Editor");
 
-        public WorkerGenerationJob(string baseOutputDir, IFileSystem fileSystem, DetailsStore detailsStore, bool force)
-            : base(baseOutputDir, fileSystem, detailsStore, force)
+        public WorkerGenerationJob(CodegenJobOptions options, IFileSystem fileSystem, DetailsStore detailsStore)
+            : base(options.AsEditor(), fileSystem, detailsStore)
```
Comment (paulbalaji) [inline workers/unity/Packages/io.improbable.gdk.buildsystem/.codegen/Source/WorkerGenerationJob.cs:22]: I quite like this way of using `.AsEditor` to signify it's for editor code

File: workers/unity/Packages/io.improbable.gdk.tools/GdkToolsConfigurationWindow.cs (hunk 2 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/io.improbable.gdk.tools/GdkToolsConfigurationWindow.cs:19]: These are rendered under the "Code Generation" header in the settings, what did you have in mind? 

image

File: workers/unity/Packages/io.improbable.gdk.buildsystem/.codegen/Source/WorkerGenerationJob.cs (hunk 3 above)
Comment (jamiebrynes7) [reply workers/unity/Packages/io.improbable.gdk.buildsystem/.codegen/Source/WorkerGenerationJob.cs:22]: It was either that or an attribute, and I didn't feel like writing more reflection code 😂

File: workers/unity/Packages/io.improbable.gdk.tools/GdkToolsConfigurationWindow.cs (hunk 2 above)
Comment (paulbalaji) [reply workers/unity/Packages/io.improbable.gdk.tools/GdkToolsConfigurationWindow.cs:19]: ah yes I didn't expand to see the heading 

ignore this
